polkadot/node/tracking-allocator/src/lib.rs

```diff
@@ -0,0 +1,226 @@
+	lock: &'a Spinlock<T>,
+}
+
+// SAFETY: Data under `UnsafeCell<T>` are protected by an exclusive lock, so they may be shared
+// between threads safely
+unsafe impl<T: Send> Sync for Spinlock<T> {}
+
+impl<T: Send> Spinlock<T> {
+	pub const fn new(t: T) -> Spinlock<T> {
+		Spinlock { lock: AtomicBool::new(false), data: UnsafeCell::new(t) }
+	}
+
+	#[inline]
+	pub fn lock(&self) -> SpinlockGuard<T> {
+		loop {
+			// Try to acquire the lock.
+			if self
+				.lock
+				.compare_exchange_weak(false, true, Ordering::Acquire, Ordering::Relaxed)
+				.is_ok()
+			{
+				return SpinlockGuard { lock: self }
+			}
+			// We failed to acquire the lock; wait until it's unlocked.
+			//
+			// In theory this should result in less coherency traffic as unlike `compare_exchange`
+			// it is a read-only operation, so multiple cores can execute it simultaneously
+			// without taking an exclusive lock over the cache line.
+			while self.lock.load(Ordering::Relaxed) {
+				std::hint::spin_loop();
+			}
+		}
+	}
+
+	#[inline]
+	fn unlock(&self) {
+		self.lock.store(false, Ordering::Release);
+	}
+}
+
+impl<T: Send> Deref for SpinlockGuard<'_, T> {
+	type Target = T;
+
+	fn deref(&self) -> &T {
+		// SAFETY: It is safe to dereference a guard to the `UnsafeCell` underlying data as the
+		// presence of the guard means the data are already locked.
+		unsafe { &*self.lock.data.get() }
+	}
+}
+
+impl<T: Send> DerefMut for SpinlockGuard<'_, T> {
+	fn deref_mut(&mut self) -> &mut T {
+		unsafe { &mut *self.lock.data.get() }
+	}
+}
+
+impl<T: Send> Drop for SpinlockGuard<'_, T> {
+	fn drop(&mut self) {
+		self.lock.unlock();
+	}
+}
+
+struct TrackingAllocatorData {
+	current: isize,
+	peak: isize,
+	limit: isize,
+	failure_handler: Option<Box<dyn Fn() + Send>>,
+}
+
+impl TrackingAllocatorData {
+	fn start_tracking(
+		mut guard: SpinlockGuard<Self>,
+		limit: isize,
+		failure_handler: Option<Box<dyn Fn() + Send>>,
+	) {
+		guard.current = 0;
+		guard.peak = 0;
+		guard.limit = limit;
+		// Cannot drop it yet, as it would trigger a deallocation
+		let old_handler = guard.failure_handler.take();
+		guard.failure_handler = failure_handler;
+		drop(guard);
+		drop(old_handler);
+	}
+
+	fn end_tracking(mut guard: SpinlockGuard<Self>) -> isize {
+		let peak = guard.peak;
+		guard.limit = 0;
+		// Cannot drop it yet, as it would trigger a deallocation
+		let old_handler = guard.failure_handler.take();
+		drop(guard);
+		drop(old_handler);
+		peak
+	}
+
+	#[inline]
+	fn track(mut guard: SpinlockGuard<Self>, alloc: isize) -> Option<SpinlockGuard<Self>> {
+		guard.current += alloc;
+		if guard.current > guard.peak {
+			guard.peak = guard.current;
+		}
+		if guard.limit == 0 || guard.peak <= guard.limit {
+			None
+		} else {
+			Some(guard)
+		}
+	}
+}
+
+static ALLOCATOR_DATA: Spinlock<TrackingAllocatorData> =
+	Spinlock::new(TrackingAllocatorData { current: 0, peak: 0, limit: 0, failure_handler: None });
+
+pub struct TrackingAllocator<A: GlobalAlloc>(pub A);
```

Comment:
The struct here is public (so there can be multiple of these), but the tracking data is static.

While this is not a problem in practice right now, it is kinda suboptimal API design I think.

It'd probably be better to e.g. only export a macro or something that would allow the user to define a global allocator without being able to make as many `TrackingAllocator` structs as they want. Maybe something like this:

```rust
macro_rules! define_tracking_allocator {
    (#[global_allocator] static $name:ident: $kind:ty = $init:expr;) => {
        #[global_allocator]
        static $name: $crate::TrackingAllocator<$kind> = $crate::TrackingAllocator::__internal_new($init);
    }
}
```

So the user would use this:

```rust
define_tracking_allocator! {
    #[global_allocator]
    static ALLOC: tikv_jemallocator::Jemalloc = tikv_jemallocator::Jemalloc;
}
```

Then the fact that there can only be a single global allocator would prevent multiple `TrackingAllocator`s from existing, hence the fact that there can be only one `ALLOCATOR_DATA` couldn't become a problem.

Reply:
Hmmm, I think I'm not following the idea. The item referred to in a macro must still be visible from the invocation site, so `TrackingAllocator` and `__internal_new` must be `pub`, and that couldn't prevent the second instantiation of `TrackingAllocator`.

I agree the API shouldn't allow for multiple instances, but I still need to figure out how to achieve that.